python310Packages.jax: 0.3.6 -> 0.3.16

mcwitt opened the pull request:
###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

Upgrades `jax` to [0.3.16](https://github.com/google/jax/blob/main/CHANGELOG.md#jax-0316) and `jaxlib` to [0.3.15](https://github.com/google/jax/blob/main/CHANGELOG.md#jaxlib-0315-july-22-2022).

~Still a draft because this depends on https://github.com/NixOS/nixpkgs/pull/183052.~

~**Note**: This marks the darwin build as broken because `jax>=0.3.14` sets `macos_minimum_os=10.14`, but as far as I can tell 10.12 is the latest supported for x86_64 in nixpkgs (relevant issue: https://github.com/NixOS/nixpkgs/issues/101229).~

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

mcwitt commented on pkgs/development/python-modules/jax/default.nix:
```diff
@@ -29,31 +30,21 @@ buildPythonPackage rec {
 
-  patches = [
-    # See https://github.com/google/jax/issues/7944
-    ./cache-fix.patch
```
The linked issue is not closed, but the patch no longer applies and the caching code seems to have changed significantly, so I'm leaning toward thinking this is no longer an issue.

mcwitt replied:
Update: linked issue is closed now

mcwitt commented on pkgs/development/python-modules/jaxlib/default.nix:
```diff
@@ -156,7 +163,7 @@ let
     # Copy-paste from TF derivation.
     # Most of these are not really used in jaxlib compilation but it's simpler to keep it
     # 'as is' so that it's more compatible with TF derivation.
```
Per the comment, I synced `TF_SYSTEM_LIBS` with the current TF derivation

mcwitt commented on pkgs/development/python-modules/jaxlib/default.nix:
```diff
@@ -165,14 +172,14 @@ let
       "com_github_grpc_grpc"
-      "com_google_protobuf"
+      # Fails with the error: /build/output/external/com_google_protobuf/BUILD.bazel:50:8: in cmd attribute of genrule rule @com_google_protobuf//:link_proto_files: $(PROTOBUF_INCLUDE_PATH) not defined
+      # "com_google_protobuf"
```
I wasn't able to get this working with protobuf in nixpkgs, even after trying several variations on the method in the [tensorflow derivation](https://github.com/NixOS/nixpkgs/blob/9da9722e432e40d877e07cdee44e896ae426f874/pkgs/development/python-modules/tensorflow/default.nix#L286-L289).

samuela replied:
That's a bummer, thanks for taking a good stab at it!

uri-canva replied:
The tensorflow derivation runs configure: https://github.com/NixOS/nixpkgs/blob/9da9722e432e40d877e07cdee44e896ae426f874/pkgs/development/python-modules/tensorflow/default.nix#L345, which picks up `PROTOBUF_INCLUDE_PATH` like this: https://sourcegraph.com/github.com/tensorflow/tensorflow@b9c969ef42b3a1172af3a79f3497aee74f3658c0/-/blob/configure.py?L1016-1018, jaxlib doesn't have the equivalent configuration logic, but we can specify `--define=PROTOBUF_INCLUDE_PATH="${protobuf-core}/include"` in the bazel flags explicitly.

uri-canva replied:
How `--define` plays into this is a bit poorly documented because its used is discouraged, having been replaced by build settings and toolchains, but make variables in general are documented here: https://bazel.build/reference/be/make-variables, and build variables assigned with `--define` is made available as a make variable.

mcwitt replied:
Thanks for the tips @uri-canva ! I'm not sure what changed, but after rebasing I was able to revert this diff (i.e. go back to using nixpkgs protobuf) without the above error.

mcwitt replied:
> Thanks for the tips @uri-canva ! I'm not sure what changed, but after rebasing I was able to revert this diff

Hmm, actually I must've been mistaken -- I started seeing the `PROTOBUF_INCLUDE_PATH` errors in the ofborg build even though I could've sworn I got past that locally.

Now I've explicitly set `PROTOBUF_INCLUDE_PATH` in the bazel flags with the method @uri-canva suggested, and that seems to have fixed it.

mcwitt commented:
Status update: I'm stuck trying to get this to build on x86_64 darwin, getting errors like the following:
```
external/org_tensorflow/tensorflow/stream_executor/allocator_stats.h:45:3: error: no template named 'optional' in namespace 'std'; did you mean 'absl::optional'?
```

I'm less familiar with darwin, but it seems like this might be an issue with nixpkgs using an unsupported darwin SDK version for x86_64. Since [v0.3.14](https://github.com/google/jax/commit/7f7358cb5807acdf679c3cb692bb0fb621dd2a52), jax has bumped `macos_minimum_os` to 10.14, but as far as I can tell nixpkgs currently only supports 10.12 for x86_64 (relevant issue: https://github.com/NixOS/nixpkgs/issues/101229).

samuela commented on pkgs/development/python-modules/jax/default.nix:
```diff
@@ -29,31 +30,23 @@ buildPythonPackage rec {
     scipy
     typing-extensions
-  ];
+  ] ++ etils.optional-dependencies.epath;
```
Do we really need to drag these optional dependencies in? Why are they necessary?

mcwitt replied:
This is because jax includes `etils[epath]` in `install_requires`: https://github.com/google/jax/blob/jax-v0.3.15/setup.py#L52

samuela commented on pkgs/development/python-modules/jaxlib/default.nix:
```diff
@@ -117,6 +123,8 @@ let
     ] ++ lib.optionals cudaSupport [
       cudatoolkit
       cudnn
+    ] ++ lib.optionals (!stdenv.isDarwin) [
+      nsync
     ];
```
I was under the impression that the Darwin build was being disabled until we get the macOS 10.14 SDKs?

mcwitt replied:
This came from syncing with the tensorflow derivation (here: https://github.com/NixOS/nixpkgs/blob/3524da2f0ab911c9c65ac6fb34e87b9698a4e5fb/pkgs/development/python-modules/tensorflow/default.nix#L232-L234)
My thought was to leave the darwin-specific workarounds in place (even though the build is disabled for now) in case switching to the newer SDK fixes the remaining issues. I'm not really sure of the recommended practice in this situation, though -- happy to remove the gating if that seems preferable.

samuela replied:
Gotcha, yeah now that we will have darwin builds working soon (shout out to @uri-canva!), we can figure how to make this work after all

samuela commented on pkgs/development/python-modules/jaxlib/default.nix:
```diff
@@ -197,7 +203,9 @@ let
       "zlib"
-    ];
+    ] ++ lib.optionals (!stdenv.isDarwin) [
+      "nsync" # fails to build on darwin
```
Does this mean that the package fails to build on darwin entirely, or that it fails to build without this particular line of code? Or does it mean that the vendored version of nsync fails to build?

mcwitt replied:
I think it means that nsync in nixpkgs is not supported on darwin (darwin is not listed in platforms: https://github.com/NixOS/nixpkgs/blob/603db2cc147884acd25d0eb02cbf536e68c56bfb/pkgs/development/libraries/nsync/default.nix#L25-L27))

(This was taken from the tensorflow derivation.)

samuela reviewed (commented):
Thanks for putting this together @mcwitt! Overall this looks thorough and most of the changes make sense to me. Just a few questions here and there...

uri-canva commented:
The darwin build might be ok after https://github.com/NixOS/nixpkgs/pull/184395, since that updates the package to use SDK 11.0, which is available in nixpkgs, it's just not the default one.

samuela commented:
> The darwin build might be ok after https://github.com/NixOS/nixpkgs/pull/184395, since that updates the package to use SDK 11.0, which is available in nixpkgs, it's just not the default one.

Once we get https://github.com/NixOS/nixpkgs/pull/184395 merged, we should def rebase this on top of that

mcwitt committed:
python310Packages.jax: 0.3.6 -> 0.3.16

python310Packages.jaxlib: 0.3.0 -> 0.3.15

mcwitt commented:
It looks like the ofborg build is failing with

```
error: building of '/nix/store/mb0mnbqh2n6lh55krxahirdcqfn06crw-python3.10-jax-0.3.15.drv!out' from .drv file timed out after 3600 seconds
```

Is there a way to increase this timeout?

samuela commented:
> Once we get #184395 merged, we should def rebase this on top of that

now merged!

onny commented:
jax 0.3.16 is available now https://github.com/google/jax/releases/tag/jax-v0.3.16

mcwitt commented:
> now merged! @mcwitt could you please rebase on top of that and then we can get this PR moving?

Just FYI, I'll be traveling for a few days and won't have a chance to work on this again before next week. If anyone needs it sooner, feel free to take over! Otherwise I should have a chance to rebase and upgrade to 0.3.16 next week.

SuperSandro2000 commented on pkgs/development/python-modules/jax/default.nix:
```diff
@@ -29,31 +30,23 @@ buildPythonPackage rec {
-    ./cache-fix.patch
-
-    # See https://github.com/google/jax/issues/10292
-    (fetchpatch {
```
Please remove the unused fetchpatch from inputs.

SuperSandro2000 commented on pkgs/development/python-modules/jax/default.nix:
```diff
@@ -29,31 +30,23 @@ buildPythonPackage rec {
   # CPU wheel is packaged.
   propagatedBuildInputs = [
     absl-py
+    (etils.overrideAttrs (_: {
+      doInstallCheck = false; # needed to break dependency cycle
+    }))
```
Please use this https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/python-modules/pylama/default.nix#L70-L72 instead.

mcwitt replied:
Actually, both seem unnecessary as there is no dependency cycle (since the etils tests are disabled). This is removed now; we can address the dependency cycle when enabling the etils tests (see https://github.com/NixOS/nixpkgs/pull/183051#issuecomment-1224954563)

SuperSandro2000 commented on pkgs/development/python-modules/jaxlib/default.nix:
```diff
@@ -48,9 +51,10 @@ let
+  version = "0.3.15";
 
   meta = with lib; {
+    broken = stdenv.isDarwin;
```
broke goes last in meta. It being first was an automated accident.

mcwitt replied:
Fixed by rebase

samuela commented:
> Just FYI, I'll be traveling for a few days and won't have a chance to work on this again before next week. If anyone needs it sooner, feel free to take over! Otherwise I should have a chance to rebase and upgrade to 0.3.16 next week.

No worries, safe travels @mcwitt! There's no rush AFAIK

samuela commented on pkgs/development/python-modules/jax/default.nix:
```diff
@@ -17,97 +17,96 @@
+    };
   };
-}
+in
+jax
```
Please use `let usingMKL = ...; in buildPythonPackage { ... }` instead of `let usingMKL = ...; jax = buildPythonPackage { ... }; in jax`

samuela replied:
Edit: I'm guessing that this is necessary for `passthru.tests`... I would still much prefer to break the dependency cycle at the etils level since that package is much, much smaller than JAX.

samuela replied:
Also passthru.tests does not get run as reliably as the standard test suite in most tooling

mcwitt replied:
Should be resolved now (see https://github.com/NixOS/nixpkgs/pull/183051#issuecomment-1224954563)

samuela commented on pkgs/development/python-modules/jax/default.nix:
```diff
@@ -17,97 +17,96 @@
-    scipy
-    typing-extensions
-  ];
+    # tests have dependency cycle: jax -> etils -> jax
+    doCheck = false;
```
We should really keep the jax test suite running. That may mean breaking the dependency cycle at etils, but the JAX test suite is comprehensive and important to get right.

mcwitt replied:
Resolved (https://github.com/NixOS/nixpkgs/pull/183051#issuecomment-1224954563)

samuela reviewed (changes requested):
This etils dependency cycle is bad news, but I would really really prefer that we break that dependency cycle at the etils derivation instead of here. A few reasons:
* etils is far smaller, less relied on.
* JAX has a large test suite that is very important
* changing the JAX derivation just with whitespace changes ruins the git blame. OTOH etils is a much smaller and younger derivation

mcwitt commented:
> Update: looks like jax is not a required dependency of etils anyhow?

@samuela correct, but jax is a test dependency of etils: https://github.com/NixOS/nixpkgs/blob/1b74710c69c2ba3b693ebae342a4bfdfd7a0c1e5/pkgs/development/python-modules/etils/default.nix#L75

This is not actually an issue right now as the etils tests are currently disabled (until https://github.com/NixOS/nixpkgs/pull/186690 is merged). So maybe best to remove the workaround here and instead break the cycle in etils when we enable tests there?

I went ahead and reverted the change that disabled checks and added to `passthru.tests`. (In case it's helpful when reviewing, the unsquashed branch is here: https://github.com/mcwitt/nixpkgs/tree/upgrade-jax)

mcwitt commented on pkgs/development/python-modules/jax/default.nix:
```diff
@@ -29,31 +30,23 @@ buildPythonPackage rec {
-    ./cache-fix.patch
-
-    # See https://github.com/google/jax/issues/10292
-    (fetchpatch {
```
Thanks, done!

samuela reviewed (approved):
Ok, this LGTM now... I'll try running nixpkgs-review and then I think we should be good to go

samuela commented:
Result of `nixpkgs-review pr 183051` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>32 packages failed to build:</summary>
  <ul>
    <li>python310Packages.aeppl</li>
    <li>python310Packages.aesara</li>
    <li>python310Packages.chex</li>
    <li>python310Packages.dalle-mini</li>
    <li>python310Packages.distrax</li>
    <li>python310Packages.dm-haiku</li>
    <li>python310Packages.dm-sonnet</li>
    <li>python310Packages.elegy</li>
    <li>python310Packages.flax</li>
    <li>python310Packages.jmp</li>
    <li>python310Packages.optax</li>
    <li>python310Packages.pymc</li>
    <li>python310Packages.rlax</li>
    <li>python310Packages.tensorflow-datasets</li>
    <li>python310Packages.treex</li>
    <li>python310Packages.vqgan-jax</li>
    <li>python39Packages.aeppl</li>
    <li>python39Packages.aesara</li>
    <li>python39Packages.chex</li>
    <li>python39Packages.dalle-mini</li>
    <li>python39Packages.distrax</li>
    <li>python39Packages.dm-haiku</li>
    <li>python39Packages.dm-sonnet</li>
    <li>python39Packages.elegy</li>
    <li>python39Packages.flax</li>
    <li>python39Packages.jmp</li>
    <li>python39Packages.optax</li>
    <li>python39Packages.pymc</li>
    <li>python39Packages.rlax</li>
    <li>python39Packages.tensorflow-datasets</li>
    <li>python39Packages.treex</li>
    <li>python39Packages.vqgan-jax</li>
  </ul>
</details>
<details>
  <summary>16 packages built:</summary>
  <ul>
    <li>python310Packages.arviz</li>
    <li>python310Packages.augmax</li>
    <li>python310Packages.jax</li>
    <li>python310Packages.jaxlib (python310Packages.jaxlib-build ,python310Packages.jaxlibWithoutCuda)</li>
    <li>python310Packages.jaxlibWithCuda</li>
    <li>python310Packages.numpyro</li>
    <li>python310Packages.objax</li>
    <li>python310Packages.treeo</li>
    <li>python39Packages.arviz</li>
    <li>python39Packages.augmax</li>
    <li>python39Packages.jax</li>
    <li>python39Packages.jaxlib (python39Packages.jaxlib-build ,python39Packages.jaxlibWithoutCuda)</li>
    <li>python39Packages.jaxlibWithCuda</li>
    <li>python39Packages.numpyro</li>
    <li>python39Packages.objax</li>
    <li>python39Packages.treeo</li>
  </ul>
</details>

mcwitt commented:
> Result of nixpkgs-review pr 183051 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)

I skimmed over the breakages. Many seem to be caused by the following:

- chex: broken by upstream removal of `tree_multimap`. Fixed by upgrading to latest (0.1.4)
- jmp: broken by upstream removal of `tree_multimap`. Fixed by upgrading to latest (https://github.com/deepmind/jmp/commit/260e5ba01f46b10c579a61393e6c7e546aeae93e)
- distrax: many tests fail; upgrade to latest not possible since it adds a `numpy<1.23` version bound
- tensorflow-datasets: long-running test suite; not sure if it'll be fixed by the above fixes or a version bump

The others seem to be broken by unrelated issues (mostly incompatible `scipy` and `rich` versions).

mcwitt commented:
Unfortunately rebasing with https://github.com/NixOS/nixpkgs/pull/184395 merged hasn't magically fixed the darwin build.

- ~Bazel fails to find `libtool`.~ I think this is fixed by adding `cctools` to `nativeBuildInputs` (https://github.com/mcwitt/nixpkgs/commit/4f8786c6d02dd1690ee74d72610c4c9ceb9ec23a)
- `ERROR: /private/tmp/nix-build-bazel-build-jaxlib-0.3.15.drv-0/output/external/llvm-project/mlir/BUILD.bazel:6549:11: Compiling mlir/lib/TableGen/Argument.cpp failed: undeclared inclusion(s) in rule '@llvm-project//mlir:TableGen'` ([log](https://logs.nix.ci/?key=nixos/nixpkgs.183051&attempt_id=fe336da2-c899-425e-b4fc-a0e9328d45b0)). @uri-canva do you have any insight here?

uri-canva commented:
I have not seen the `undeclared inclusion(s) in rule` in my testing. Did you test if jaxlib works on its own without this branch? Maybe it's already broken in master.

uri-canva commented:
It shows as successful 9 hours ago: https://hydra.nixos.org/eval/1777705?filter=jaxlib&compare=1777668&full=

uri-canva commented:
Oh I see, it's probably some new transitive dependency from the version upgrade.

uri-canva commented:
Seems to be a known issue with bazel:
https://github.com/tensorflow/tensorflow/issues/55563
https://github.com/bazelbuild/bazel/issues/15359

uri-canva commented:
The workarounds might be a bit hard for us though, not sure how well gcc works on darwin, but we can try.

mcwitt commented:
> Just to confirm: @mcwitt are you building on x86_64-darwin?

@samuela yes, the `undeclared inclusion(s)` error occurs for x86_64-darwin. Here's the ofborg build log: https://logs.nix.ci/?key=nixos/nixpkgs.183051&attempt_id=fe336da2-c899-425e-b4fc-a0e9328d45b0

samuela commented:
Is there any hope trying to fix or patch the x86_64-darwin bug ourselves? It appears to be a quite hairy issue in bazel IIUC... Perhaps we should also raise it with the JAX folks?

mcwitt commented:
> I'm ok letting these slide for now as long as someone promises to come back and do the necessary upgrades after this PR is merged.

@samuela I'd be happy to handle upgrading chex and jmp after this is merged (already have the updates building in local branches so it shouldn't be too much work 🤞 )

samuela commented:
@mcwitt @uri-canva How do ya'll feel about merging this as is, despite the fact that it breaks the x86_64-darwin build? On the one hand I hate to break the build... OTOH a) this breakage isn't our fault, b) I don't want PRs to get stuck forever. wdyt?

uri-canva commented:
Yeah I think it's ok, I only just fixed the darwin build, and I'm not sure how long it was broken before that, so I don't think there would be anyone depending on it.

uri-canva commented:
Let's mark it as broken and add a comment with all the context / links to the issues for future reference. That way when someone looks at it in the future they won't have to start figuring it out from scratch.

samuela commented:
Yeah, I def don't want you to think that the x86_64-darwin work was for nothing @uri-canva! We should def commit to getting darwin support back up and running as soon as upstream progress on this issue allows.

SuperSandro2000 commented:
so we can merge this now or not?

samuela commented:
I think we can!
